tensordict/prototype/tensorclass.py

```diff
@@ -481,7 +483,26 @@ def _setitem(self, item: NestedKey, value: Any) -> None:
         isinstance(item, tuple) and all(isinstance(_item, str) for _item in item)
     ):
         raise ValueError("Invalid indexing arguments.")
-    if is_tensorclass(value) and not isinstance(value, self.__class__):
+
+    if not is_tensorclass(value) and not isinstance(
+        value, (TensorDictBase, numbers.Number, Tensor, MemmapTensor)
+    ):
+        raise ValueError(
+            f"__setitem__ only supports tensorclasses, tensordicts,"
+            f" numeric scalars and tensors. Got {type(value)}"
+        )
+
+    # if it is one of accepted "broadcast" types
+    if not is_tensorclass(value):
+        # attempt broadcast on all tensordata and nested tensorclasses
+        self._tensordict[item] = value
+        for key, val in self._non_tensordict.items():
+            if is_tensorclass(val):
+                _setitem(self._non_tensordict[key], item, value)
+        return
+
+    # else it is tensorclass
+    if not isinstance(value, self.__class__):
```

Comment:
Perhaps this can be moved inside `else` block under above `if` statement and merged with `else` block below. I.e.

```python
if not is_tensorclass(value):
    ...
else:
    if not isinstance(value, self.__class__):
        # check keys
    # contents of else block below
```

Could even consider swapping these blocks so the initial statement is `if is_tensorclass(value)`. I find minimising negations makes things easier to read back.

Reply:
I agree on negation, I chose to keep it in this case to avoid a long nested if that will also decrease readability. 
But I can swap if you prefer so.

Reply:
didn't notice we can get rid off of the return. definitely worth swapping then